### PR TITLE
[CPDLP-1532] Declare after clawback

### PR DESCRIPTION
**Pull request by asmega**
### Context

- Ticket: https://dfedigital.atlassian.net/browse/CPDLP-1532
- Declaring a clawback already works this adds some tests to confirm it

### Changes proposed in this pull request

- Add test to confirm re-declaring after a clawback
- Refactored out some shared tests
- Refactored tests so i think they are more comprehendible

### Guidance to review

- none

### CODE REVIEW

File: app/services/record_declarations/retained/ecf.rb [hunk 1]
```diff
@@ -3,12 +3,14 @@
...
-      include RecordDeclarations::ECF
-
       class << self
         def valid_evidence_types
           %w[training-event-attended self-study-material-completed other].freeze
         end
+
+        def declaration_model
+          ParticipantDeclaration::ECF
+        end
```
Comment (StupidCodeFactory) [inline app/services/record_declarations/retained/ecf.rb:13]: Could we pass this as this an argument to the service class?

File: spec/services/record_declarations/started/early_career_teacher_spec.rb [hunk 2]
```diff
@@ -2,51 +2,290 @@
...
+        },
+      )
+    end
+
+    it "raises an error" do
+      expect { subject.call }.to raise_error(ActionController::ParameterMissing, /The property '#\/participant_id' must be a valid Participant ID/)
+    end
+  end
+
+  context "happy path" do
```
Comment (StupidCodeFactory) [inline spec/services/record_declarations/started/early_career_teacher_spec.rb:90]: Can we find a better description please?

File: spec/services/record_declarations/started/early_career_teacher_spec.rb [hunk 3]
```diff
@@ -2,51 +2,290 @@
...
+          cpd_lead_provider:,
+          declaration_date:,
+          declaration_type:,
+          evidence_held: "foo",
+        },
+      )
+    end
+
+    it "ignores the superfluous parameter" do
+      expect { subject.call }.to change { ParticipantDeclaration.count }.by(1)
```
Comment (StupidCodeFactory) [inline spec/services/record_declarations/started/early_career_teacher_spec.rb:57]:
```suggestion
      expect { subject.call }.to change { profile.participant_declarations.where(declaration_date:, course_identifier:, declaration_type:, evidence_held: nil, declaration_date:).count }.by(1)
```

File: spec/services/record_declarations/started/early_career_teacher_spec.rb [hunk 4]
```diff
@@ -2,51 +2,290 @@
...
+
+      new_declaration = ParticipantDeclaration.order(created_at: :asc).last
+
+      expect(new_declaration.evidence_held).to be_nil
```
Comment (StupidCodeFactory) [inline spec/services/record_declarations/started/early_career_teacher_spec.rb:61]: ✂️

File: spec/services/record_declarations/started/early_career_teacher_spec.rb [hunk 5]
```diff
@@ -2,51 +2,290 @@
...
+    end
+  end
+
+  context "happy path" do
+    before do
+      Induction::Enrol.new(induction_programme:, participant_profile: profile).call
+    end
+
+    it "creates the declaration" do
+      expect { subject.call }.to change { ParticipantDeclaration.count }.by(1)
```
Comment (StupidCodeFactory) [inline spec/services/record_declarations/started/early_career_teacher_spec.rb:96]: same a above, be a bit more specific in the query.

File: spec/services/record_declarations/started/early_career_teacher_spec.rb [hunk 6]
```diff
@@ -2,51 +2,290 @@
...
     end
   end
 
-  context "when user is for 2020 cohort" do
-    let!(:cohort_2020) { create(:cohort, start_year: 2020) }
-    let!(:school_cohort_2020) { create(:school_cohort, cohort: cohort_2020, school: ect_profile.school) }
+  context "when user profile ParticipantProfileState is withdrawn and also withdrawn on declaration date" do
+    before do
+      Induction::Enrol.new(induction_programme:, participant_profile: profile).call
+      create(:participant_profile_state, :withdrawn, participant_profile: profile, cpd_lead_provider:, created_at: declaration_date_object - 2.days)
```
Comment (StupidCodeFactory) [inline spec/services/record_declarations/started/early_career_teacher_spec.rb:187]: can you use the service class + travel_to to do this ?

File: spec/services/record_declarations/started/early_career_teacher_spec.rb [hunk 7]
```diff
@@ -2,51 +2,290 @@
...
+        subject.call
+      }.to raise_error(ActionController::ParameterMissing, /Declaration must be before withdrawal date/)
+       .and not_change { ParticipantDeclaration.count }
+    end
+  end
+
+  context "when user profile ParticipantProfileState is deferred, but was active on declaration date" do
+    before do
+      Induction::Enrol.new(induction_programme:, participant_profile: profile).call
+      create(:participant_profile_state, :deferred, participant_profile: profile, cpd_lead_provider:)
```
Comment (StupidCodeFactory) [inline spec/services/record_declarations/started/early_career_teacher_spec.rb:201]: same here, please use the service class

File: spec/services/record_declarations/started/early_career_teacher_spec.rb [hunk 8]
```diff
@@ -2,51 +2,290 @@
...
+
+    it "does not create the declaration" do
+      expect {
+        subject.call
+      }.to raise_error(ActionController::ParameterMissing, /The property '#\/course_identifier' must be an available course to '#\/participant_id'/)
+       .and not_change { ParticipantDeclaration.count }
+    end
+  end
+
+  context "when user is for 2020 cohort (ie NQT+1)" do
```
Comment (StupidCodeFactory) [inline spec/services/record_declarations/started/early_career_teacher_spec.rb:239]: Do we even need the anymore? i.e are we still receiving declarations for NQT+1 (could very well be)

Comment (asmega) [reply spec/services/record_declarations/started/early_career_teacher_spec.rb:239]: my understanding is we should never receive a declaration for an NQT+1. and if we do we should never create a declaration as no one should ever get paid for these

File: app/services/record_declarations/retained/ecf.rb (hunk 1 above)
Comment (asmega) [reply app/services/record_declarations/retained/ecf.rb:13]: sure but not for this change. I don't want to modify the API, i'm merely inlining code from elsewhere so the change is minimal

File: spec/services/record_declarations/started/early_career_teacher_spec.rb (hunk 6 above)
Comment (asmega) [reply spec/services/record_declarations/started/early_career_teacher_spec.rb:187]: which service class are you referring to?

File: spec/services/record_declarations/started/early_career_teacher_spec.rb (hunk 4 above)
Comment (asmega) [reply spec/services/record_declarations/started/early_career_teacher_spec.rb]: done

File: spec/services/record_declarations/started/early_career_teacher_spec.rb (hunk 3 above)
Comment (asmega) [reply spec/services/record_declarations/started/early_career_teacher_spec.rb]: done

File: spec/services/record_declarations/started/early_career_teacher_spec.rb (hunk 6 above)
Comment (StupidCodeFactory) [reply spec/services/record_declarations/started/early_career_teacher_spec.rb:205]: `Participants::Withdraw::(ECT|Mentor)`

File: spec/services/record_declarations/started/early_career_teacher_spec.rb (hunk 5 above)
Comment (asmega) [reply spec/services/record_declarations/started/early_career_teacher_spec.rb]: done

File: spec/services/record_declarations/started/early_career_teacher_spec.rb (hunk 2 above)
Comment (asmega) [reply spec/services/record_declarations/started/early_career_teacher_spec.rb]: done

File: spec/services/record_declarations/started/early_career_teacher_spec.rb (hunk 6 above)
Comment (asmega) [reply spec/services/record_declarations/started/early_career_teacher_spec.rb]: done

File: spec/services/record_declarations/started/early_career_teacher_spec.rb (hunk 7 above)
Comment (asmega) [reply spec/services/record_declarations/started/early_career_teacher_spec.rb:237]: done

**Review by leoapost — APPROVED** [review]

**Pull request merged**